### PR TITLE
ChainerX dynamic op registry

### LINKED ISSUE #6674: OpRegistry, its Backend interface, and op registration macro

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #6674

### CODE REVIEW

**Comment by niboshi** [comment]
TODO: 
* Test (registration to backends' instance op registry)
* CMakeList.txt

File: chainerx_cc/chainerx/backend.h [hunk 1]
```diff
@@ -40,6 +49,17 @@ class Backend {
     // Queries if the backend supports data transfer between two devices.
     virtual bool SupportsTransfer(Device& src_device, Device& dst_device) = 0;
 
+    // Calls the op implementation.
+    template <typename OpType, typename... Args>
+    auto CallOp(Args&&... args) {
+        Op& op = op_registry_.GetOp<OpType>();
+        return dynamic_cast<OpType*>(&op)->Call(std::forward<Args>(args)...);
```
Comment (niboshi) [inline chainerx_cc/chainerx/backend.h]: TODO: cast to `OpType&`

File: chainerx_cc/chainerx/backend.h [hunk 2]
```diff
@@ -21,6 +24,9 @@ class Backend {
     Backend& operator=(const Backend&) = delete;
     Backend& operator=(Backend&&) = delete;
 
+    // Initializes the backend instantce.
+    void Initialize();
```
Comment (niboshi) [inline chainerx_cc/chainerx/backend.h]: This could be virtual.

File: chainerx_cc/chainerx/backend_testdata/backend0.cc [hunk 3]
```diff
@@ -11,6 +12,12 @@ class Backend0 : public chainerx::native::NativeBackend {
     using NativeBackend::NativeBackend;
 
     std::string GetName() const override { return "backend0"; }
+
+protected:
+    chainerx::OpRegistry& GetParentOpRegistry() {
+        static chainerx::OpRegistry op_registry{&chainerx::native::NativeBackend::GetGlobalOpRegistry()};
```
Comment (niboshi) [inline chainerx_cc/chainerx/backend_testdata/backend0.cc]: Use heap

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
@niboshi Could you fix this? https://travis-ci.org/chainer/chainer/jobs/513958913#L398

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by niboshi** [comment]
Fixed clang-tidy errors.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins failure is unrelated to this PR. LGTM.

**Pull request merged**